### DISCUSSION TITLE
Multiple token input with past gives nonsense output logits

### DISCUSSION THREAD

**Discussion by yansongzhou** [QwenLM/Qwen3.6#84] [Q&A] [answered]
When I pass multiple tokens in input_ids along with past_key_values (the qwen3.5 dynamic cache), the output logits is nonsense.
I have no problem processing tokens with past one by one, but when I pass multiple tokens in a single forward pass, the result is not the same as when I handle single tokens in a loop.
I suspect that this has something to do with the new architecture, because the same code works (in the case of multiple tokens with past) for qwen3-vl. Is this not supported or am I doing it wrong?

## is this wrong or not supported?
    forward_kwargs = {
        "input_ids": current_input_ids, # when it's (batch_size, x>1) the output is nonsense; when it's (batch_size, 1) everything's fine
        "past_key_values": past_key_values, # the past key values from the previous forward
        "return_dict": True,
        "use_cache": True,
    }
    with torch.inference_mode():
        outputs = self.model(**forward_kwargs)

ps I tried adjusting some reasonable parameters like attention_mask and position_ids, but they don't seem to make any difference. The single token loop just works and the multiple tokens pass just doesn't.

**Comment by yansongzhou** [comment]
I did more experiments and proposed my answer: this may just be unsupported.
The first reason, as suggested in some issues and conversations, is numerical errors. Even without a past, passing several tokens at once and iterating give slightly different results. This may mainly be caused by non-transformer architectures, as this model clearly uses recurrent states.
Furthermore when I pass multiple new tokens with past, the model seems to forget most of its context in past_key_values. This may be amplified by the update of recursive hidden states, which is not as clearly separated as tokens in transformer modules. I suppose during training the model either sees an initial sequence of tokens without a past or generate one new token at a time, so it may not be common for multiple tokens to be passed with a past, and the model just doesn't know what to do with that.
Anyways I could be terribly wrong, which is in fact good news for me as my project rely on generation of multiple tokens. Any help is appreciated.

**Comment by yansongzhou** [comment]
So I read the source code and this might be the root cause:
https://github.com/huggingface/transformers/blob/d64a6d67d8c004a25570db4df5689e06caea6af7/src/transformers/models/qwen3_5/modular_qwen3_5.py#L345
The recurrent_state is reinitialized in this case, causing loss of context.
My current fix is reusing the recurrent_states (but not the conv states). It kinda works in my case but I'm not really sure.

**Comment by yansongzhou** [comment]
ps: This fix does not guarantee the same output as when you run the generation token by token. It just gives a reasonable response, without loss of context.

**Comment by JJJYmmm** [comment, marked as answer]
@yansongzhou  Hello! I think you should remove the `seqlen == 1` check here if you want to pass multiple tokens at one time.  (transformers doesn't support mtp or chunked profill currently, so here is just the safety check) https://github.com/huggingface/transformers/blob/d64a6d67d8c004a25570db4df5689e06caea6af7/src/transformers/models/qwen3_5/modular_qwen3_5.py#L275

**Reply by yansongzhou** [reply]
I just turned to running GatedDeltaNet in a loop for multiple token inputs, but this is way better.
Many thanks.
